torch/legacy/optim/aggmo.py

```diff
@@ -0,0 +1,84 @@
+import torch
+
+
+def aggmo(opfunc, x, config, state=None):
+    """A plain implementation of AggMo
+
+    ARGS:
+
+    - `opfunc` : a function that takes a single input (X), the point
+                of a evaluation, and returns f(X) and df/dX
+    - `x`      : the initial point
+    - `config` : a table with configuration parameters for the optimizer
+    - `config['learningRate']`      : learning rate
+    - `config['learningRateDecay']` : learning rate decay
+    - `config['weightDecay']`       : weight decay
+    - `config['weightDecays']`      : vector of individual weight decays
+    - `config['momentum']`          : momentum
+    - `config['learningRates']`     : vector of individual learning rates
+    - `state`  : a table describing the state of the optimizer; after each
+                call the state is modified
+    - `state['evalCounter']`        : evaluation counter (optional: 0, by default)
+
+    RETURN:
+    - `x`     : the new x vector
+    - `f(x)`  : the function, evaluated before the update
+
+    (Clement Farabet, 2012)
+    """
+    # (0) get/update state
+    state = state if state is not None else config
+    lr = config.get('learningRate', 1e-3)
+    lrd = config.get('learningRateDecay', 0)
+    wd = config.get('weightDecay', 0)
+    mom = config.get('momentum', [0.0, 0.9, 0.99])
+    lrs = config.get('learningRates', None)
+    wds = config.get('weightDecays', None)
+    if 'evalCounter' not in state:
+        state['evalCounter'] = 0
+    if wd != 0 and wds is not None:
+        raise ValueError("Only one of wd and wds can be specified")
+
+    # (1) evaluate f(x) and df/dx
+    fx, dfdx = opfunc(x)
+
+    # (2) weight decay with single or individual parameters
+    if wd != 0:
+        dfdx.add_(wd, x)
+    elif wds is not None:
+        if not state['decayParameters']:
+            state['decayParameters'] = torch.Tensor().type_as(x).resize_as_(dfdx)
+
+        state['decayParameters'].copy_(wds).mul_(x)
+        dfdx.add_(state['decayParameters'])
+
+    # (3) apply momentum
+    dx = torch.zeros_like(dfdx)
+    if len(mom) != 0:
+        if 'dfdx' not in state:
+            state['dfdx'] = {}
+            for b in mom:
+                state['dfdx'][b] = torch.zeros_like(dfdx)
+        for b in mom:
+            state['dfdx'][b].mul_(b).add_(dfdx)
```

Comment:
Hi @AtheMathmo, Firstly, great paper, it was super fun to read! Somehow I ended up here and am skimming through this code... Wondering why you don't add -dfdx and then omit the negative sign on the clr variable below.

From what I see you're doing:
Vt = beta*Vt-1 + grad(f(x))
x = x - gamma*VT

Reply:
I did this only to be consistent with the code here: https://github.com/pytorch/pytorch/blob/master/torch/legacy/optim/sgd.py#L62-L84

I'm not sure if there is any deeper meaning to this!